README.md
# needle-retry
[![npm version](http://img.shields.io/npm/v/needle-retry.svg)](https://www.npmjs.org/package/needle-retry)
[![Build Status](http://img.shields.io/travis/alexlangberg/node-needle-retry.svg)](https://travis-ci.org/alexlangberg/node-needle-retry)
[![Dependency Status](https://david-dm.org/alexlangberg/node-needle-retry.svg)](https://david-dm.org/alexlangberg/node-needle-retry)
[![devDependency Status](https://david-dm.org/alexlangberg/node-needle-retry/dev-status.svg)](https://david-dm.org/alexlangberg/node-needle-retry#info=devDependencies)

Combines [needle](https://www.npmjs.org/package/needle) with [retry](https://www.npmjs.org/package/retry) to account for unstable webservers. Thus, needle-retry will retry for a certain number of times (default 5) before giving up. Currently implements ```request``` and ```get```. Tests currently only cover ```GET``` requests but everything should work too, as this is basically just a wrapper.

## Installation
```
npm install needle-retry
```

## Options
Options can be passed in as the second parameter, as an object with a property ```needle``` for needle options and a property ```retry``` for retry options. This module has a single option, ```fullDocument``` that will make it retry if a full HTML document is not received. Defaults to ```false```. Default options:

```javascript
var options = {
    needleRetry: {
        fullDocument: false
    },
    needle: {
        follow_max: 20
    },
    retry: {
        retries: 5
    }
}
```

Have a look at the respective doc pages for [needle](https://www.npmjs.org/package/needle) and [retry](https://www.npmjs.org/package/retry) for a list of available options.

## Example
```javascript
var needle = require('needle-retry');

needle.get('http://www.github.com', function(error, response) {
  console.log(response.body);
});
```

## Example request
```javascript
var needle = require('needle-retry');

var options = {
    needleRetry: {
        fullDocument: false
    },
    needle: {
        follow_max: 20
    },
    retry: {
        retries: 5
    }
}

var data = {
    foo: 'bar'
}

needle.request('get', 'http://www.github.com', data, options, function(error, response) {
  console.log(response.body);
});
```


CHANGELOG.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
# Changelog

- [v1.0.0 (2015-10-11)](#v100-2015-10-11)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

### v1.0.0 (2015-10-11)

 * First release.


test/test.js
'use strict';

var test = require('tape');
var http = require('http');
var R = require('ramda');
var nr = require('../');

var serverUrl = 'http://localhost:1337';
var url = serverUrl + '/200';
var server;

var defaultOptions = {
  'retry': {
    'minTimeout': 0,
    'maxTimeout': 0
  }
};

test('setup', function(t) {
  var failures = 0;
  server = http.createServer(function(request, response) {
    var code = request.url.substring(1);

    if (request.url === '/200') {
      response.writeHead(code, {'Content-Type': 'text/plain'});
      response.end('<p>Hello World</p>');
    }

    if (request.url === '/301' || request.url === '/302') {
      response.writeHead(code, {'Content-Type': 'text/plain'});
      response.end('<p>Not found</p>');
    }

    if (request.url === '/404') {
      response.writeHead(code, {'Content-Type': 'text/plain'});
      response.end('<p>Not found</p>');
    }

    if (request.url === '/thirdtimeacharm') {
      if (failures < 2) {
        failures++;
        response.writeHead(500, {'Content-Type': 'text/plain'});
        response.end('<p>Server error!</p>');
      } else {
        response.writeHead(200, {'Content-Type': 'text/plain'});
        response.end('<p>Hello World</p>');
      }
    }

    if (request.url === '/onlyfulldocument') {
      response.writeHead(200, {'Content-Type': 'text/plain'});
      response.end('<html><body><p>Incomplete</p>');
    }
  }).listen(1337);

  server.on('listening', function() {
    t.end();
  });
});

test('it loads without options', function(t) {
  nr.get(url, function() {
    t.end();
  });
});

test('it loads with options', function(t) {
  nr.get(url, defaultOptions, function() {
    t.end();
  });
});

test('it can request', function(t) {
  t.plan(3);

  nr.request('get', url, null, defaultOptions, function(error, response, body) {
    t.equal(error, null);
    t.equal(response.statusCode, 200);
    t.equal(body, '<p>Hello World</p>');
  });
});

test('it can request with get', function(t) {
  t.plan(3);

  nr.get(url, defaultOptions, function(error, response, body) {
    t.equal(error, null);
    t.equal(response.statusCode, 200);
    t.equal(body, '<p>Hello World</p>');
  });
});

test('it handles errors from needle', function(t) {
  t.plan(1);

  nr.get('foo', defaultOptions, function(error) {
    t.equal(error.errno, 'ENOTFOUND');
  });
});

test('it throws on too many redirects', function(t) {
  var url = serverUrl + '/301';
  var options = R.merge(defaultOptions, {
    'needle': {
      'follow_max': 0
    }
  });

  t.plan(2);

  nr.get(url, options, function(error, response) {
    t.equal(response.statusCode, 301);
    t.equal(error.message, 'Too many redirects. Increase in needle options.');
  });
});

test('it throws on 404', function(t) {
  var url = serverUrl + '/404';

  t.plan(2);

  nr.get(url, defaultOptions, function(error, response) {
    t.equal(response.statusCode, 404);
    t.equal(error.message, 'Request failed. Status code: 404.');
  });
});

test('it throws on incomplete document option', function(t) {
  var url = serverUrl + '/onlyfulldocument';
  var options = R.merge(defaultOptions, {
    'needleRetry': {
      'fullDocument': true
    }
  });

  t.plan(2);

  nr.get(url, options, function(error, response) {
    t.equal(response.statusCode, 200);
    t.equal(error.message, 'fullDocument: not full document.');
  });
});

test('it can retry on failure', function(t) {
  t.plan(3);

  nr.get(
    serverUrl + '/thirdtimeacharm',
    defaultOptions,
    function(error, response, body) {
      t.equal(error, null);
      t.equal(response.statusCode, 200);
      t.equal(body, '<p>Hello World</p>');
    });
});

test('teardown', function(t) {
  server.close();
  t.end();
});


lib/index.js
'use strict';

var R = require('ramda');
var needle = require('needle');
var retry = require('retry');
var nr;

function getDefaults(options) {
  var defaults = {
    'needleRetry': {
      'fullDocument': false
    },
    'needle': {
      'follow_max': 20
    },
    'retry': {
      'retries': 5
    }
  };

  if (!options) {
    return defaults;
  }

  return {
    'needle': R.merge(defaults.needle, options.needle),
    'retry': R.merge(defaults.retry, options.retry),
    'needleRetry': R.merge(
      defaults.needleRetry,
      options.needleRetry
    )
  };
}

function request(method, url, data, options, callback) {
  needle.request(method, url, data, options.needle, function(error, response, body) {
    if (error) {
      return callback(error);
    }

    if (response.statusCode === 301 || response.statusCode === 302) {
      return callback(
        new Error('Too many redirects. Increase in needle options.'),
        response
      );
    }

    if (response.statusCode !== 200) {
      return callback(
        new Error('Request failed. Status code: ' + response.statusCode + '.'),
        response
      );
    }

    if (options.needleRetry.fullDocument) {
      if (body.indexOf('</html>') === -1) {
        return callback(
          new Error('fullDocument: not full document.'),
          response,
          body
        );
      }
    }

    return callback(null, response, body);
  });
}

function run(method, url, data, userOptions, callback) {
  var operation;
  var options;
  if (typeof userOptions === 'function') {
    callback = userOptions;
    options = getDefaults();
  } else {
    options = getDefaults(userOptions);
  }

  operation = retry.operation(options.retry);
  operation.attempt(function() {
    request(method, url, data, options, function(error, response, body) {
      if (operation.retry(error)) {
        return undefined;
      }

      return callback(
        error ? operation.mainError() : null,
        response,
        body
      );
    });
  });
}

nr = {
  'get': function(url, options, callback) {
    return run('get', url, null, options, callback);
  },
  'request': function(method, url, data, options, callback) {
    return run(method, url, data, options, callback);
  }
};

module.exports = nr;
